### DISCUSSION TITLE
I have a question on a Route that is ~10% longer than a shorter one

### DISCUSSION THREAD

**Discussion by eugen-nw** [valhalla/valhalla#4277] [Q&A]
I wonder if anyone can help me on this. I'm really new to Routing, so please bear with me. We're calling into a Valhalla OMS server that we'd setup on a VM. We're requesting Routes using "costing": "truck","directions_options":{"units":"miles"} as we're computing Routes for trucks. The input locations are below
```
{
    "Locations": [
        {
            "Identifier": "OKC",
            "Coordinate": {
                "Latitude": 35.456573,
                "Longitude": -97.604025
            }
        },
        {
            "Identifier": "LIT",
            "Coordinate": {
                "Latitude": 34.731471,
                "Longitude": -92.252331
            }
        },
        {
            "Identifier": "SMF",
            "Coordinate": {
                "Latitude": 38.654953,
                "Longitude": -121.48726
            }
        }
    ]
}
```
The direct Route returned LIT -> SMF = 2152.13 miles.
If we assemble the Route manually, this indirect Route gives a shorter total distance: LIT -> OKC -> SMF = 356.184 + 1634.58 = 1990.764 miles.

What options / configuration settings could affect this behavior please?

**Comment by eugen-nw** [comment]
The request we send is source_to_targets.  I've just noticed in the documentation the optimized_routes, should I try using that?  I cannot try it out right now because I'm away from my work computer.

**Comment by eugen-nw** [comment]
I've got to play a bit with `optimized_routes`.  It behaves quite differently from `sources_to_targets`: it takes as input the `locations` array and builds _one_ optimized Route out of those Coordonates.  Whereas `sources_to_targets` takes the two `sources` and `targets` arrays and builds up the entire matrix of Routes in between the elements of the two arrays.

By seeing that an `optimized_routes` call with an array of only two Coordinates takes 2+ seconds on the valhalla_server, I would assume that it would take forever to have a `source_to_targets` call in which every single Route is optimized.  It's worth mentioning that our Valhalla server VM is quite small in terms of CPU and memory as we're still just getting our feet wet in this area.

Please correct me if I'm wrong.

**Reply by nilsnolde** [reply]
Did you look at the docs? http://valhalla.github.io/valhalla/api/optimized/api-reference/#

You want to optimize waypoints. Valhalla can only do intermediate waypoints and has generally a pretty poor solver for waypoint optimization. If you need smth more sophisticated or that can do start/end/intermediate, have a look at VROOM, it plays fairly nice with Valhalla.

**Reply by eugen-nw** [reply]
Thanks very much @nilsnolde!  Yes, I'm reading the docs and have a steep learning curve.  On the topic of waypoints, we've noticed that Valhalla picked a longer direct route from A -> C than the alternative sum of A -> B and B -> C on a different road.  Is this related to your suggestion please?

**Reply by nilsnolde** [reply]
That could have all kinds of reasons related to restrictions, sub-optimality of the bidirectional search or probably others.

**Reply by nilsnolde** [reply]
Also please remember that the default is not distance optimization, but cost takes many more things into account other the length! You can turn that on with shortest=true though. I’d recommend spending a few days to understand the subject. Or alternatively contact me on enquiry@gis-ops.com for commercial support.

**Reply by eugen-nw** [reply]
Thanks very much @nilsnolde again!  Totally makes sense, let me learn more about this topic.  Where should I look for info on costs for trucks please?

Where would the shortest=true setting go please?  I'd guess in the ~/src/valhalla/valhalla.json file.  Perhaps in the "service_limits" -> "truck" section?

**Reply by nilsnolde** [reply]
nope, it's a request parameter.
docs: https://valhalla.github.io/valhalla/api/turn-by-turn/api-reference/
config options: https://github.com/valhalla/valhalla/blob/master/scripts/valhalla_build_config

**Reply by eugen-nw** [reply]
On the first attempt I experienced no difference in the results using the shortest=true option :-(  Is there anything missing from the request that I send please?

Translating the original JSON into the sources_to_targets body below returns the same distance & time regardless of the presence or absence of the "shortest" option:
```{"sources":[{"lat":35.456573,"lon":-97.604025},{"lat":34.731471,"lon":-92.252331},{"lat":38.654953,"lon":-121.48726}],"targets":[{"lat":35.456573,"lon":-97.604025},{"lat":34.731471,"lon":-92.252331},{"lat":38.654953,"lon":-121.48726}],"costing": "truck", "shortest": "true","directions_options":{"units":"miles"}}```

The interesting part of the result does not change either way.  My concern is that the 0 -> 2 distance would be 356.184 + 1634.580 = 1,990.764 should the 1 -> 0 -> 2 route been taken.
```
{
    "distance": 2152.13,
    "time": 107523,
    "to_index": 2,
    "from_index": 1
}
```

**Reply by eugen-nw** [reply]
I experimented a bit more with this.  There may be some good reasons why Valhalla decided to route the trucks on the longer route.  I replaced "truck" with "auto" in "costing" and got the desired shorter distance.  Even without using the "shortest" : "true" option.  That's the route that Google Maps returns as well.
```
{
    "distance": 1976.2,
    "time": 100026,
    "to_index": 2,
    "from_index": 1
}
```
